src/faebryk/core/cpp/src/pathfinder/bfs.cpp

```diff
@@ -0,0 +1,182 @@
+/* This file is part of the faebryk project
+ * SPDX-License-Identifier: MIT
+ */
+
+#include "pathfinder/bfs.hpp"
+#include "perf.hpp"
+#include <deque>
+#include <sstream>
+
+std::string Edge::str() const {
+    std::stringstream ss;
+    ss << from->get_full_name(false) << "->" << to->get_full_name(false);
+    return ss.str();
+}
+
+std::string PathStackElement::str() /*const*/ {
+    std::stringstream ss;
+    if (up) {
+        ss << child_type.get_name() << "->" << parent_type.get_name() << "." << name;
+    } else {
+        ss << parent_type.get_name() << "." << name << "->" << child_type.get_name();
+    }
+    return ss.str();
+}
+
+bool UnresolvedStackElement::match(PathStackElement &other) {
+    return elem.parent_type == other.parent_type &&
+           elem.child_type == other.child_type && elem.name == other.name &&
+           elem.up != other.up;
+}
+
+std::string UnresolvedStackElement::str() /*const*/ {
+    std::stringstream ss;
+    ss << elem.str();
+    if (split) {
+        ss << " split";
+    }
+    return ss.str();
+}
+
+// BFSPath implementations
+BFSPath::BFSPath(/*const*/ GI_ref_weak path_head)
+  : Path(path_head)
+  , path_data(std::make_shared<PathData>()) {
+}
+
+BFSPath::BFSPath(const BFSPath &other)
+  : Path(other)
+  , path_data(std::make_shared<PathData>(*other.path_data))
+  , confidence(other.confidence)
+  , filtered(other.filtered)
+  , stop(other.stop) {
+}
+
+BFSPath::BFSPath(const BFSPath &other, /*const*/ GI_ref_weak new_head)
+  : Path(other)
+  , path_data(other.path_data)
+  , confidence(other.confidence)
+  , filtered(other.filtered)
+  , stop(other.stop) {
+    path.push_back(new_head);
+    assert(!other.filtered);
+}
+
+BFSPath::BFSPath(BFSPath &&other)
+  : Path(std::move(other))
+  , path_data(std::move(other.path_data))
+  , confidence(other.confidence)
+  , filtered(other.filtered)
+  , stop(other.stop) {
+}
+
+BFSPath BFSPath::operator+(/*const*/ GI_ref_weak gif) {
+    return BFSPath(*this, gif);
+}
+
+PathData &BFSPath::get_path_data_mut() {
+    if (!path_data.unique()) {
```

Comment:
Deprecated method causing build errors on modern compilers: https://en.cppreference.com/w/cpp/memory/shared_ptr/unique

Obtaining file:///Users/mattwildoer/Projects/faebryk
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Installing backend dependencies ... done
  Preparing editable metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Preparing editable metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [60 lines of output]
      *** scikit-build-core 0.9.2 using CMake 3.27.8 (wheel)
      *** Configuring CMake...
      2024-11-12 11:24:30,939 - scikit_build_core - WARNING - libdir/ldlibrary: /opt/homebrew/opt/python@3.12/Frameworks/Python.framework/Versions/3.12/lib/Python.framework/Versions/3.1
2/Python is not a real file!                                                                                                                                                                   loading initial cache file /private/var/folders/8v/zd4990r94sj86c3cz5gszwcc0000gn/T/tmpdptqu5gm/build/CMakeInit.txt
      -- The CXX compiler identification is AppleClang 15.0.0.15000309
      -- Detecting CXX compiler ABI info
      -- Detecting CXX compiler ABI info - done
      -- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ - skipped
      -- Detecting CXX compile features
      -- Detecting CXX compile features - done
      -- Found Python: /Users/mattwildoer/Projects/faebryk/.venv/bin/python3.12 (found version "3.12.7") found components: Interpreter Development.Module
      -- Python_EXECUTABLE: /Users/mattwildoer/Projects/faebryk/.venv/bin/python3.12
      -- Python_INCLUDE_DIR: /opt/homebrew/Cellar/python@3.12/3.12.7_1/Frameworks/Python.framework/Versions/3.12/include/python3.12
      -- Python_LIBRARY: /opt/homebrew/opt/python@3.12/Frameworks/Python.framework/Versions/3.12/Python
      -- Configuring done (14.5s)
      -- Generating done (0.0s)
      -- Build files have been written to: /private/var/folders/8v/zd4990r94sj86c3cz5gszwcc0000gn/T/tmpdptqu5gm/build
      *** Building project with Ninja...
      [1/24] Building CXX object CMakeFiles/faebryk_core_cpp.dir/src/pathfinder/bfs.cpp.o
      FAILED: CMakeFiles/faebryk_core_cpp.dir/src/pathfinder/bfs.cpp.o
      /Library/Developer/CommandLineTools/usr/bin/c++ -DEDITABLE=0 -Dfaebryk_core_cpp_EXPORTS -I/Users/mattwildoer/Projects/faebryk/src/faebryk/core/cpp/include -I/opt/homebrew/Cellar/p
ython@3.12/3.12.7_1/Frameworks/Python.framework/Versions/3.12/include/python3.12 -I/private/var/folders/8v/zd4990r94sj86c3cz5gszwcc0000gn/T/pip-build-env-jorra5vc/overlay/lib/python3.12/site-packages/nanobind/include -O2 -O3 -DNDEBUG -std=gnu++20 -arch x86_64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX15.1.sdk -fPIC -fvisibility=hidden -fno-stack-protector -Os -MD -MT CMakeFiles/faebryk_core_cpp.dir/src/pathfinder/bfs.cpp.o -MF CMakeFiles/faebryk_core_cpp.dir/src/pathfinder/bfs.cpp.o.d -o CMakeFiles/faebryk_core_cpp.dir/src/pathfinder/bfs.cpp.o -c /Users/mattwildoer/Projects/faebryk/src/faebryk/core/cpp/src/pathfinder/bfs.cpp                                                                                                   /Users/mattwildoer/Projects/faebryk/src/faebryk/core/cpp/src/pathfinder/bfs.cpp:78:20: error: no member named 'unique' in 'std::shared_ptr<PathData>'
          if (!path_data.unique()) {
               ~~~~~~~~~ ^
      1 error generated.
      [2/24] Building CXX object CMakeFiles/faebryk_core_cpp.dir/src/pathfinder/pathcounter.cpp.o
      [3/24] Building CXX object CMakeFiles/faebryk_core_cpp.dir/src/graph/link.cpp.o
      [4/24] Building CXX object CMakeFiles/faebryk_core_cpp.dir/src/graph/path.cpp.o
      [5/24] Building CXX object CMakeFiles/faebryk_core_cpp.dir/src/graph/links.cpp.o
      [6/24] Building CXX object CMakeFiles/faebryk_core_cpp.dir/src/graph/graphinterfaces.cpp.o
      [7/24] Building CXX object CMakeFiles/faebryk_core_cpp.dir/src/graph/graphinterface.cpp.o
      [8/24] Building CXX object CMakeFiles/faebryk_core_cpp.dir/src/graph/node.cpp.o
      [9/24] Building CXX object CMakeFiles/faebryk_core_cpp.dir/src/graph/graph.cpp.o
      [10/24] Building CXX object CMakeFiles/faebryk_core_cpp.dir/src/main.cpp.o
      ninja: build stopped: subcommand failed.
      Traceback (most recent call last):
        File "/Users/mattwildoer/Projects/faebryk/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/Users/mattwildoer/Projects/faebryk/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/Users/mattwildoer/Projects/faebryk/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 176, in prepare_metadata_for_build_edi
table                                                                                                                                                                                              whl_basename = build_hook(metadata_directory, config_settings)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/8v/zd4990r94sj86c3cz5gszwcc0000gn/T/pip-build-env-jorra5vc/overlay/lib/python3.12/site-packages/hatchling/build.py", line 83, in build_editable
          return os.path.basename(next(builder.build(directory=wheel_directory, versions=['editable'])))
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/8v/zd4990r94sj86c3cz5gszwcc0000gn/T/pip-build-env-jorra5vc/overlay/lib/python3.12/site-packages/hatchling/builders/plugin/interface.py", line 147, in 
build                                                                                                                                                                                              build_hook.initialize(version, build_data)
        File "/private/var/folders/8v/zd4990r94sj86c3cz5gszwcc0000gn/T/pip-build-env-jorra5vc/overlay/lib/python3.12/site-packages/scikit_build_core/hatch/plugin.py", line 121, in initi
alize                                                                                                                                                                                              self._initialize(build_data=build_data)
        File "/private/var/folders/8v/zd4990r94sj86c3cz5gszwcc0000gn/T/pip-build-env-jorra5vc/overlay/lib/python3.12/site-packages/scikit_build_core/hatch/plugin.py", line 237, in _init
ialize                                                                                                                                                                                             builder.build(build_args=build_args)
        File "/private/var/folders/8v/zd4990r94sj86c3cz5gszwcc0000gn/T/pip-build-env-jorra5vc/overlay/lib/python3.12/site-packages/scikit_build_core/builder/builder.py", line 229, in bu
ild                                                                                                                                                                                                self.config.build(
        File "/private/var/folders/8v/zd4990r94sj86c3cz5gszwcc0000gn/T/pip-build-env-jorra5vc/overlay/lib/python3.12/site-packages/scikit_build_core/cmake.py", line 243, in build
          self._build(*local_args, *build_args)
        File "/private/var/folders/8v/zd4990r94sj86c3cz5gszwcc0000gn/T/pip-build-env-jorra5vc/overlay/lib/python3.12/site-packages/scikit_build_core/cmake.py", line 254, in _build
          raise FailedLiveProcessError(msg) from None
      scikit_build_core.errors.FailedLiveProcessError: CMake build failed
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.

[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: pip install --upgrade pip
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.